### PR TITLE
Migrate Guava cache usage to Caffeine

### PR DESCRIPTION
**Pull request by Stephan202**
This PR is a replacement of #286, which was closed as a side effect of the deletion of its base branch. For the original context, see #258. This migration from Guava to Caffeine was unlocked by the migration to Java 8 in the context of #276.

I'll re-apply the comments from the other PR, so as to ease the review effort.

### CODE REVIEW

File: agent-bridge/src/main/java/com/newrelic/agent/bridge/ExtensionHolder.java
Comment (Stephan202) [inline agent-bridge/src/main/java/com/newrelic/agent/bridge/ExtensionHolder.java:21]: I found only one implementation of this interface. If it _is_ used elsewhere we should revert this change and introduce a `try/catch` construct in the implementation.

File: newrelic-agent/build.gradle
Comment (Stephan202) [inline newrelic-agent/build.gradle:97]: Caffeine 3.x requires Java 11+; the 2.x line will [remain supported](https://github.com/ben-manes/caffeine/releases/tag/v3.0.0).

File: newrelic-agent/src/main/java/com/newrelic/agent/ThreadService.java
Comment (Stephan202) [inline newrelic-agent/src/main/java/com/newrelic/agent/ThreadService.java:83]: Throughout the PR: `ExecutionException` isn't thrown anymore, so I dropped associated `try/catch` blocks where reasonable, taking into account whether the backing code could actually throw an exception and whether the fallback logic "made sense" or seemed to be there just to appease the compiler.

File: newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java
Comment (Stephan202) [inline newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java:40]: For context see the discussion @ https://github.com/newrelic/newrelic-java-agent/pull/286#discussion_r621944669. We should consider dropping this line.

File: newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java
Comment (Stephan202) [inline newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java:42]: This `Runnable::run` is here because one of the tests expects to see the effects of the removal notification in a timely manner. See the [wiki](https://github.com/ben-manes/caffeine/wiki/Guava) on how Caffeine may schedule this kind of work on another thread. Alternatively the test could be tweaked. If the current behavior is desired, perhaps `.executor(Runnable::run)` should also be added in `AsyncTransactionService`.

File: newrelic-agent/src/main/java/com/newrelic/agent/extension/ExtensionRewriter.java
Comment (Stephan202) [inline newrelic-agent/src/main/java/com/newrelic/agent/extension/ExtensionRewriter.java:30]: I haven't investigated what this code actually does. Does it make sense to add Caffeine, or should it replace Guava, or...?

File: newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/weaver/ClassLoaderClassTransformer.java
Comment (Stephan202) [inline newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/weaver/ClassLoaderClassTransformer.java:131]: `GuavaBackedExtensionClass` is now Caffeine-backend, so a rename seemed in order.

File: newrelic-agent/src/main/java/com/newrelic/agent/tracers/metricname/MetricNameFormats.java
Comment (Stephan202) [inline newrelic-agent/src/main/java/com/newrelic/agent/tracers/metricname/MetricNameFormats.java:67]: Note that the `if` condition is now only evaluated on cache misses.

File: newrelic-agent/src/test/java/com/newrelic/agent/extension/DependencyRemapperTest.java
Comment (Stephan202) [inline newrelic-agent/src/test/java/com/newrelic/agent/extension/DependencyRemapperTest.java:90]: Related to my question above: no idea whether these test changes make sense.

File: newrelic-weaver/src/test/java/com/newrelic/weave/weavepackage/WeavePackageManagerTest.java
Comment (Stephan202) [inline newrelic-weaver/src/test/java/com/newrelic/weave/weavepackage/WeavePackageManagerTest.java:242]: See the [wiki](https://github.com/ben-manes/caffeine/wiki/Guava); Caffeine's size-based eviction may cause the cache to grow _slightly_ larger than configured. Here we trigger an explicit cleanup to make sure the test passes. Under normal operation this kind of cleanup is performed in the context of subsequent cache operations.

**Review by Stephan202 — COMMENTED** [review]
Re-applied the comments from #286.

File: newrelic-weaver/src/test/java/com/newrelic/weave/weavepackage/WeavePackageManagerTest.java
Comment (Stephan202) [reply newrelic-weaver/src/test/java/com/newrelic/weave/weavepackage/WeavePackageManagerTest.java]: This one `.cleanUp()` call wasn't enough; added a commit with a more robust fix.

**Comment by XiXiaPdx** [comment]
@Stephan202  Just a heads up that we are planning to review this after releasing agent version 7.0.0, which we are targeting for June 7th.  Thank you for your contribution and patience!

**Comment by Stephan202** [comment]
No hurry :)

For when the time comes: google/guava#5571 provides additional background.

**Comment by tbradellis** [comment]
@Stephan202  I've started doing some testing for this PR.  I'm curious if this build with Caffeine resolves the deadlock we were seeing in the support ticket where discussing this change to Caffeine originally came up.  Are you able to run this agent build in that environment?

**Comment by Stephan202** [comment]
@tbradellis so far I did not attempt to deploy this code. Given our setup this wouldn't be impossible per se, but also a bit more work than I can commit to right now. I'll make a note in case I do find some spare time. :crossed_fingers:

(For the time being we're still running with `-Dnewrelic.config.class_transformer.com.newrelic.instrumentation.java.completable-future-jdk8u40.enabled=false`. Haven't tried dropping it to see whether we're still hitting the issue in the first place. Will make a note to schedule such a test.)

**Comment by tbradellis** [comment]
Hey @Stephan202 . I'm working to get this merged.  I need to rebase and I also have a few minor changes for some of the comments.  
Can you give [permission on the fork](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork) for me to push changes?

**Comment by Stephan202** [comment]
Hi @tbradellis. It [turns out](https://github.community/t/how-can-we-enable-allow-edits-from-maintainers-by-default/2847) that the "Allow edits from maintainers" feature isn't available for PRs created from organization accounts (such as `PicnicSupermarket`). Instead I have invited you as an "outside collaborator" on the [repository](https://github.com/PicnicSupermarket/newrelic-java-agent) itself.

If you have any questions about this or the code, let me know. Tnx for picking this up!

**Comment by Stephan202** [comment]
I've rebased the branch on latest `main`; there were no conflicts.

W.r.t. open tasks, cursory inspection surfaces two more:
- There are a few hits for `git grep -i 'guava cache'`; some/all of those should now probably reference Caffeine.
- If `THIRD_PARTY_NOTICES.md` is manually maintained (seems so :eyes:) it should also be updated.

**Comment by Stephan202** [comment]
It's not clear to me whether/how the Scala test failures relate to the changes in this PR.

**Comment by tbradellis** [comment]
> Scala test failures

We have an issue with out scala tests that we are working to fix.  It's also a race condition so, once it reruns a few times it'll pass.

**Comment by tbradellis** [comment]
@Stephan202 It looks like I stepped on your commits  when I had to --reset-author to get my license/cla accepted.  I'll correct that.

**Comment by kford-newrelic** [comment]
Hi @Stephan202 check your email when you have a moment (and your junk folder just in case) I tried to send you a message

**Comment by Stephan202** [comment]
@kford-newrelic it took me a while to find that email, as a GMail filter got hold of it; replied :)

@tbradellis I took the liberty to fix the author situation in this PR. (I still had an old version of the branch, so simply cherry-picked your commit on top of it and then rebased; there were no conflicts and the final diff is identical.)

File: newrelic-agent/src/main/java/com/newrelic/agent/ThreadService.java
Comment (tbradellis) [reply newrelic-agent/src/main/java/com/newrelic/agent/ThreadService.java:85]: This works for us

File: newrelic-agent/src/main/java/com/newrelic/agent/extension/ExtensionRewriter.java
Comment (tbradellis) [reply newrelic-agent/src/main/java/com/newrelic/agent/extension/ExtensionRewriter.java:30]: We will leave this change in as is.

File: agent-bridge/src/main/java/com/newrelic/agent/bridge/ExtensionHolder.java
Comment (tbradellis) [reply agent-bridge/src/main/java/com/newrelic/agent/bridge/ExtensionHolder.java:21]: keep this change.

File: newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java
Comment (tbradellis) [reply newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java:40]: We are going to keep this change as is with capacity set to 8.  I'm creating a research spike to unblock getting this merged and investigate whether resizing makes sense.

**Review by tbradellis — APPROVED** [review]

**Pull request merged**
